File: server/src/daemon-api.ts
import { createServer, IncomingMessage, ServerResponse } from 'http';
import { CallManager, CallConflictError, CallForbiddenError } from './phone-call.js';
import { randomBytes } from 'crypto';

interface ClientInfo {
  clientId: string;
  connectedAt: number;
  lastHeartbeat: number;
}

interface DaemonApiConfig {
  callManager: CallManager;
  onRefCountZero: () => void;
  onRefCountPositive: () => void;
}

export class DaemonApi {
  private clients = new Map<string, ClientInfo>();
  private server: ReturnType<typeof createServer> | null = null;
  private config: DaemonApiConfig;
  private heartbeatCheckInterval: ReturnType<typeof setInterval> | null = null;
  private startTime = Date.now();

  constructor(config: DaemonApiConfig) {
    this.config = config;
  }

  start(port: number): Promise<void> {
    return new Promise((resolve) => {
      this.server = createServer((req, res) => this.handleRequest(req, res));
      this.server.listen(port, '127.0.0.1', () => {
        console.error(`[daemon] Control API listening on 127.0.0.1:${port}`);
        resolve();
      });

      // Check for dead clients every 5 seconds
      this.heartbeatCheckInterval = setInterval(() => this.checkDeadClients(), 5000);

      // Start initial shutdown timer in case no client ever connects
      this.config.onRefCountZero();
    });
  }

  shutdown(): void {
    if (this.heartbeatCheckInterval) clearInterval(this.heartbeatCheckInterval);
    this.server?.close();
  }

  private async handleRequest(req: IncomingMessage, res: ServerResponse): Promise<void> {
    const url = new URL(req.url!, `http://${req.headers.host}`);
    const method = req.method || 'GET';

    try {
      if (method === 'GET' && url.pathname === '/status') {
        return this.handleStatus(res);
      }
      if (method === 'POST' && url.pathname === '/connect') {
        return this.handleConnect(res);
      }
      if (method === 'POST' && url.pathname === '/disconnect') {
        return await this.handleDisconnect(req, res);
      }
      if (method === 'POST' && url.pathname === '/heartbeat') {
        return await this.handleHeartbeat(req, res);
      }
      if (method === 'POST' && url.pathname === '/calls') {
        return await this.handleInitiateCall(req, res);
      }

      const callMatch = url.pathname.match(/^\/calls\/([^/]+)\/(continue|speak|end)$/);
      if (method === 'POST' && callMatch) {
        const [, callId, action] = callMatch;
        return await this.handleCallAction(req, res, callId, action);
      }

      res.writeHead(404);
      res.end(JSON.stringify({ error: 'Not found' }));
    } catch (error) {
      console.error('[daemon-api] Unhandled error:', error);
      if (!res.headersSent) {
        res.writeHead(500);
        res.end(JSON.stringify({ error: 'Internal server error' }));
      }
    }
  }

  private handleStatus(res: ServerResponse): void {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({
      status: 'ok',
      uptime: Math.round((Date.now() - this.startTime) / 1000),
      connectedClients: this.clients.size,
      clientIds: Array.from(this.clients.keys()),
    }));
  }

  private handleConnect(res: ServerResponse): void {
    const clientId = randomBytes(16).toString('hex');
    this.clients.set(clientId, {
      clientId,
      connectedAt: Date.now(),
      lastHeartbeat: Date.now(),
    });
    this.config.onRefCountPositive();
    console.error(`[daemon] Client connected: ${clientId} (total: ${this.clients.size})`);
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ clientId }));
  }

  private async handleDisconnect(req: IncomingMessage, res: ServerResponse): Promise<void> {
    const body = await readJsonBody(req, res);
    if (!body) return;
    const { clientId } = body as { clientId: string };
    await this.removeClient(clientId);
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ ok: true }));
  }

  private async handleHeartbeat(req: IncomingMessage, res: ServerResponse): Promise<void> {
    const body = await readJsonBody(req, res);
    if (!body) return;
    const { clientId } = body as { clientId: string };
    const client = this.clients.get(clientId);
    if (client) {
      client.lastHeartbeat = Date.now();
    }
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ ok: true }));
  }

  private async handleInitiateCall(req: IncomingMessage, res: ServerResponse): Promise<void> {
    const body = await readJsonBody(req, res);
    if (!body) return;
    const { clientId, message } = body as { clientId: string; message: string };

    if (!this.clients.has(clientId)) {
      res.writeHead(401);
      res.end(JSON.stringify({ error: 'Unknown client' }));
      return;
    }

    try {
      const result = await this.config.callManager.initiateCall(clientId, message);
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify(result));
    } catch (error) {
      if (error instanceof CallConflictError) {
        res.writeHead(409);
        res.end(JSON.stringify({ error: error.message }));
      } else {
        const msg = error instanceof Error ? error.message : 'Unknown error';
        console.error('[daemon-api] Call error:', msg);
        res.writeHead(500);
        res.end(JSON.stringify({ error: msg }));
      }
    }
  }

  private async handleCallAction(
    req: IncomingMessage, res: ServerResponse,
    callId: string, action: string
  ): Promise<void> {
    const body = await readJsonBody(req, res);
    if (!body) return;
    const { clientId, message } = body as { clientId: string; message: string };

    if (!this.clients.has(clientId)) {
      res.writeHead(401);
      res.end(JSON.stringify({ error: 'Unknown client' }));
      return;
    }

    try {
      let result: unknown;
      switch (action) {
        case 'continue':
          result = { response: await this.config.callManager.continueCall(clientId, callId, message) };
          break;
        case 'speak':
          await this.config.callManager.speakOnly(clientId, callId, message);
          result = { ok: true };
          break;
        case 'end':
          result = await this.config.callManager.endCall(clientId, callId, message);
          break;
      }
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify(result));
    } catch (error) {
      if (error instanceof CallForbiddenError) {
        res.writeHead(403);
        res.end(JSON.stringify({ error: error.message }));
      } else {
        const msg = error instanceof Error ? error.message : 'Unknown error';
        console.error('[daemon-api] Call error:', msg);
        res.writeHead(500);
        res.end(JSON.stringify({ error: msg }));
      }
    }
  }

  private async removeClient(clientId: string): Promise<void> {
    if (!this.clients.has(clientId)) return;
    await this.config.callManager.forceEndCallByClient(clientId);
    this.clients.delete(clientId);
    console.error(`[daemon] Client disconnected: ${clientId} (total: ${this.clients.size})`);
    if (this.clients.size === 0) {
      this.config.onRefCountZero();
    }
  }

  private async checkDeadClients(): Promise<void> {
    const now = Date.now();
    const deadTimeout = 10000;
    const deadClientIds: string[] = [];
    for (const [clientId, info] of this.clients) {
      if (now - info.lastHeartbeat > deadTimeout) {
        deadClientIds.push(clientId);
      }
    }
    for (const clientId of deadClientIds) {
      console.error(`[daemon] Client ${clientId} heartbeat timeout, removing`);
      await this.removeClient(clientId);
    }
  }
}

async function readJsonBody(req: IncomingMessage, res: ServerResponse): Promise<Record<string, unknown> | null> {
  return new Promise((resolve, reject) => {
    let body = '';
    let size = 0;
    const MAX_BODY_SIZE = 1024 * 1024; // 1MB
    req.on('data', (chunk) => {
      size += chunk.length;
      if (size > MAX_BODY_SIZE) {
        res.writeHead(413);
        res.end(JSON.stringify({ error: 'Request body too large' }));
        resolve(null);
        return;
      }
      body += chunk;
    });
    req.on('end', () => {
      try {
        resolve(JSON.parse(body));
      } catch {
        res.writeHead(400);
        res.end(JSON.stringify({ error: 'Invalid JSON' }));
        resolve(null);
      }
    });
    req.on('error', reject);
  });
}

File: src/callme/daemon_api.py
"""Daemon HTTP Control API.

MCP 서버들이 HTTP로 데몬을 제어한다.
"""

from __future__ import annotations

import asyncio
import logging
import os
import time
from typing import Any, Callable, Awaitable

from aiohttp import web

from .call_manager import CallConflictError, CallForbiddenError, CallManager
from .config import compute_env_hash

log = logging.getLogger("callme.api")


class DaemonApi:
    def __init__(
        self,
        call_manager: CallManager,
        on_ref_count_zero: Callable[[], Any],
        on_ref_count_positive: Callable[[], Any],
    ) -> None:
        self._call_manager = call_manager
        self._on_ref_count_zero = on_ref_count_zero
        self._on_ref_count_positive = on_ref_count_positive
        self._clients: dict[str, dict[str, Any]] = {}
        self._start_time = time.time()
        self._env_hash = compute_env_hash()
        self._runner: web.AppRunner | None = None
        self._heartbeat_task: asyncio.Task | None = None

    async def start(self, port: int) -> None:
        app = web.Application()
        app.router.add_get("/status", self._handle_status)
        app.router.add_post("/connect", self._handle_connect)
        app.router.add_post("/disconnect", self._handle_disconnect)
        app.router.add_post("/heartbeat", self._handle_heartbeat)
        app.router.add_post("/calls", self._handle_initiate_call)
        app.router.add_post("/calls/{call_id}/continue", self._handle_call_action)
        app.router.add_post("/calls/{call_id}/speak", self._handle_call_action)
        app.router.add_post("/calls/{call_id}/end", self._handle_call_action)

        self._runner = web.AppRunner(app)
        await self._runner.setup()
        site = web.TCPSite(self._runner, "127.0.0.1", port)
        await site.start()
        log.info("Control API listening on 127.0.0.1:%d", port)

        self._heartbeat_task = asyncio.create_task(self._check_dead_clients_loop())

        # Start initial shutdown timer
        self._on_ref_count_zero()

    async def shutdown(self) -> None:
        if self._heartbeat_task:
            self._heartbeat_task.cancel()
        if self._runner:
            await self._runner.cleanup()

    # ── Handlers ──────────────────────────────────────────────────

    async def _handle_status(self, request: web.Request) -> web.Response:
        return web.json_response(
            {
                "status": "ok",
                "uptime": round(time.time() - self._start_time),
                "connectedClients": len(self._clients),
                "envHash": self._env_hash,
            }
        )

    async def _handle_connect(self, request: web.Request) -> web.Response:
        client_id = os.urandom(16).hex()
        self._clients[client_id] = {
            "connectedAt": time.time(),
            "lastHeartbeat": time.time(),
        }
        self._on_ref_count_positive()
        log.info("Client connected: %s (total: %d)", client_id, len(self._clients))
        return web.json_response({"clientId": client_id})

    async def _handle_disconnect(self, request: web.Request) -> web.Response:
        body = await request.json()
        client_id = body.get("clientId", "")
        await self._remove_client(client_id)
        return web.json_response({"ok": True})

    async def _handle_heartbeat(self, request: web.Request) -> web.Response:
        body = await request.json()
        client_id = body.get("clientId", "")
        if client_id in self._clients:
            self._clients[client_id]["lastHeartbeat"] = time.time()
        return web.json_response({"ok": True})

    async def _handle_initiate_call(self, request: web.Request) -> web.Response:
        body = await request.json()
        client_id = body.get("clientId", "")
        message = body.get("message", "")
        to = body.get("to") or None

        if client_id not in self._clients:
            return web.json_response({"error": "Unknown client"}, status=401)

        try:
            result = await self._call_manager.initiate_call(client_id, message, to=to)
            return web.json_response(result)
        except CallConflictError as e:
            return web.json_response({"error": str(e)}, status=409)
        except CallForbiddenError as e:
            return web.json_response({"error": str(e)}, status=403)
        except Exception as e:
            log.exception("Call error")
            return web.json_response({"error": str(e)}, status=500)

    async def _handle_call_action(self, request: web.Request) -> web.Response:
        call_id = request.match_info["call_id"]
        action = request.path.rsplit("/", 1)[-1]
        body = await request.json()
        client_id = body.get("clientId", "")
        message = body.get("message", "")

        if client_id not in self._clients:
            return web.json_response({"error": "Unknown client"}, status=401)

        try:
            if action == "continue":
                response = await self._call_manager.continue_call(
                    client_id, call_id, message
                )
                return web.json_response({"response": response})
            elif action == "speak":
                await self._call_manager.speak_only(client_id, call_id, message)
                return web.json_response({"ok": True})
            elif action == "end":
                result = await self._call_manager.end_call(client_id, call_id, message)
                return web.json_response(result)
            else:
                return web.json_response({"error": "Unknown action"}, status=400)
        except CallForbiddenError as e:
            return web.json_response({"error": str(e)}, status=403)
        except Exception as e:
            log.exception("Call action error")
            return web.json_response({"error": str(e)}, status=500)

    # ── Client management ─────────────────────────────────────────

    async def _remove_client(self, client_id: str) -> None:
        if client_id not in self._clients:
            return
        await self._call_manager.force_end_call_by_client(client_id)
        del self._clients[client_id]
        log.info("Client disconnected: %s (total: %d)", client_id, len(self._clients))
        if not self._clients:
            self._on_ref_count_zero()

    async def _check_dead_clients_loop(self) -> None:
        try:
            while True:
                await asyncio.sleep(5)
                now = time.time()
                dead = [
                    cid
                    for cid, info in self._clients.items()
                    if now - info["lastHeartbeat"] > 10
                ]
                for cid in dead:
                    log.warning("Client %s heartbeat timeout", cid)
                    await self._remove_client(cid)
        except asyncio.CancelledError:
            pass
